go-controller/pkg/ovn/pods.go

```diff
@@ -443,47 +443,24 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) (err error) {
 			return fmt.Errorf("error creating pod network annotation: %v", err)
 		}
 
+		// On pod creation: perform the gateway routes setup before annotating
+		// the pod with the IP annotation. For egress IP: we need to clear the
+		// per-pod created SNAT, and that event handler will be triggered by the
+		// update performed below. Doing this will ensure that we won't race by
+		// having the SNAT created here, after the performed removal by the
+		// egress IP code.
+		if err := oc.addPodGatewayRoutes(pod, podIfAddrs); err != nil {
```

Comment:
I'm not sure this is the right solution. Moving this code here under if needsIP means that the routes are only added if this is a new IP assignment, where as before it would happen regardless. I think trying to modify pods.go function is a recipe for more bugs and we shoudl use proper locking/signaling/handling between egress IP handler and pod handler to ensure no race.

Reply:
No, see L463. We add this if `!needsIP`

Reply:
And again, this doesn't have anything to do with locking. The "racy-ness" of concern here is strictly in terms of the logical order of which things are executed. The egress IP code needs to be sure that the pod GR SNAT is created before the annotation is applied, before that wasn't the case.

Reply:
Actually, why do we always add the pod GW SNAT? I realize now that this is not going to work with egress IP. They will always compete for every pod update, and with the change in this PR the egress IP code will not work because we'll continuously add the pod GW SNAT for every pod update, whereas the egress IP code only removes it once the pod has its IPs assigned.

Reply:
The one case I figured was covered by always adding the pod GW SNAT was in case someone deletes the NB DB and restarts ovnkube-master. The assumption then is that we should have the gateway routes re-created, even though the pod already has the annotation. I hence added this to `syncPods`. We are thus creating the pod gateway routes whenever a pod does not have an IP assignment yet, or when we restart, always.

Reply:
Re-iterating offline discussion: Alex to look into if we could use the port cache as our signal to know addLogicalPort is finished and then egressIP knows it can remove/add the new SNAT.

Reply:
Port cache works just fine and resolves our concurrency issue with addLogicalPort. I've updated my patch